### ISSUE TITLE
optimize code

### ISSUE THREAD

**Issue by fpvandoorn** [chikinn/hanabi#24]
I was surprised that playing 1000 games with `cheater` bot takes already a surprisingly long time (~10s on my laptop). I noticed that most of the time is spend on these commands in `hanabi_classes`:

``` python
self.HandHistory.append(deepcopy(self.h))
self.progressHistory.append(deepcopy(self.progress))
```

Commenting out those two lines speeds up the executing by a factor of 5+. Is there a way to speed up this copying? Or is it possible to not keep the whole history when the bots don't need it? (Maybe by adding a flag which disables the history, or maybe there is a smarter way.)

**Comment by benzax** [comment]
Yeah, deepcopies (https://docs.python.org/2/library/copy.html) are very expensive, but are the easy way to protect current state.  However, cards and plays should never be mutated, so lists of them should only ever require shallow copies.

I will take a look, it is almost certainly safe to replace deepcopy with copy for our use cases, and that should be a bit faster.
